// myframework/lib/MailObj.js
var nodemailer = require('../../nodemailer');
var smtpTransport = require('../../nodemailer-smtp-transport');
var fs = require('fs');
var path = require('path');
var jade = require('../../jade');

module.exports = MailObj;

function MailObj(options) {
	console.log("Options +++++++++",options.username);
	this.transporter = nodemailer.createTransport(smtpTransport({		
		host: options.host,
		tls: {rejectUnauthorized: false},
		auth: {
		     user: options.username,
		     pass: options.password
		 }
	}));
}

MailObj.prototype.notify = function (mailOptions,cb) {
	
	var mailObj = this;
	
	if(mailOptions.html){
		mailOptions.html = jade.compile(mailOptions.html)(mailOptions.params)
		mailObj.sendEmail(mailOptions,cb);
	}
	else if(mailOptions.htmltemplate){
		var filePath = path.resolve(__dirname , '../../../public/templates/' , mailOptions.htmltemplate);
		mailOptions.html = jade.compile(fs.readFileSync(filePath))(mailOptions.params)
		mailObj.sendEmail(mailOptions,cb);
	}
	else if(mailOptions.jadetemplate){
		var filePath = path.resolve(__dirname , '../../../public/templates/' , mailOptions.jadetemplate);
		mailOptions.html = jade.renderFile(filePath, mailOptions.params);
		mailObj.sendEmail(mailOptions,cb);
	}
	else{
		mailObj.sendEmail(mailOptions,cb);
	}
};

MailObj.prototype.sendEmail = function (mailOptions,cb){
	this.transporter.sendMail(mailOptions, function(error, info) {
		 if (error) 
		     cb(error, null);
		 else 
		     cb(null,info);		 
	});
	cb(null, 'Mail Sent');
};
